server/testkit/tests/mtls_test.rs

```diff
@@ -0,0 +1,353 @@
+use kanidmd_testkit::{is_free_port, PORT_ALLOC};
+use std::sync::atomic::Ordering;
+
+use std::error::Error;
+use std::net::{IpAddr, Ipv6Addr, SocketAddr};
+use std::pin::Pin;
+use std::time::Duration;
+use tokio::net::{TcpListener, TcpStream};
+use tokio::sync::oneshot;
+use tokio::time;
+use tracing::{error, trace};
+
+use openssl::asn1;
+use openssl::bn;
+use openssl::ec;
+use openssl::hash;
+use openssl::nid::Nid;
+use openssl::pkey;
+use openssl::ssl::{Ssl, SslAcceptor, SslMethod, SslRef};
+use openssl::ssl::{SslConnector, SslVerifyMode};
+use openssl::x509::extension::BasicConstraints;
+use openssl::x509::extension::ExtendedKeyUsage;
+use openssl::x509::extension::KeyUsage;
+use openssl::x509::extension::SubjectAlternativeName;
+use openssl::x509::extension::SubjectKeyIdentifier;
+use openssl::x509::X509NameBuilder;
+use openssl::x509::X509;
+use tokio_openssl::SslStream;
+
+use openssl::error::ErrorStack as OpenSSLError;
+
+/// Gets an [EcGroup] for P-256
+pub fn get_group() -> Result<ec::EcGroup, OpenSSLError> {
+    ec::EcGroup::from_curve_name(Nid::X9_62_PRIME256V1)
+}
+
+fn keylog_cb(_ssl_ref: &SslRef, key: &str) {
+    trace!(?key);
+}
+
+fn build_self_signed_client_server(
```

Comment:
Perhaps `build_self_signed_client_cert` ?

Reply:
It's going to move shortly I think as I need it in the main daemon for building the certs inside Kanidm, and allowing them to be exposed to other partso fthe code.

Also it's a hybrid client and server cert? Maybe a different name needed?

Reply:
Yeah, that's more the point - it wasn't exactly clear without reading the code what it was doing, and later it's called and I thought we were starting up a server with a self-signed cert or something 😄